doggos/knowledge/rule.py

```diff
@@ -1,24 +1,77 @@
+    """
+    Class used to represent a fuzzy rule,
+    that is used in fuzzy logic systems to infer an output based on input variables.
+    It consists from an antecedent(premise) and a consequent.
+    https://en.wikipedia.org/wiki/Fuzzy_rule
+
+    Attributes
+    --------------------------------------------
+
+    __antecedent : Antecedent
+        Antecedent of the rule. Has getter and setter.
+
+    __consequent : Consequent
+        Consequent of the rule. Has getter and setter.
+
+    Methods
+    --------------------------------------------
+    output(self, features: Dict[str, float], get_firing: bool = False) -> List[MembershipDegree] or float \
+                                                    or Tuple[List[MembershipDegree] or float, MembershipDegree]:
+        Calculate firing value and output of rule, then return output and optionally firing.
+
+    Examples:
+    --------------------------------------------
+
+    """
+
     def __init__(self, antecedent: Antecedent, consequent: Consequent):
-        self.__firing_value: MembershipDegree = None
-        self.__rule_output: List[MembershipDegree] or float = None
+        """
+        Create a new rule with defined antecedent and consequent.
+
+        :param antecedent: antecedent of the rule
+        :param consequent: consequent of the rule
+        """
+
         self.__antecedent = antecedent
         self.__consequent = consequent
 
     @property
     def antecedent(self) -> Antecedent:
-        pass
+        return self.__antecedent
+
+    @antecedent.setter
+    def antecedent(self, new_antecedent: Antecedent) -> NoReturn:
+        """Sets firing value and rule output to None"""
+        self.__antecedent = new_antecedent
 
     @property
     def consequent(self) -> Consequent:
-        pass
+        return self.__consequent
+
+    @consequent.setter
+    def consequent(self, new_consequent: Consequent) -> NoReturn:
+        """Sets rule output to None"""
+        self.__consequent = new_consequent
+
+    def output(self, features: Dict[str, float], get_firing: bool = False) -> List[MembershipDegree] or float \
+                                                    or Tuple[List[MembershipDegree] or float, MembershipDegree]:
+        """
+        Method that is calculating firing value and output of the rule.
 
-    def output(self, features: Dict[str, float]) -> List[MembershipDegree] or float:
-        pass
+        :param get_firing: an optional argument. If True, it will return also firing value of rule.
+        :param features: a dictionary with names of Linguistic Variables and elements of their domains.
+        :return: if get_firing is False - consequent output, that is tuple of list of membership degrees or one float,
+        if get_firing is True - tuple of consequent output and membership degree.
+        """
+        firing = self.__antecedent.firing(features)
+        if get_firing:
+            return self.__consequent.output(firing), firing
```

Comment:
Different parameters will need to be provided for takagi_sugeno_consequent (features)

Reply:
It has been changed.